### PR TITLE
Hash: apply performant defaults for PGA

### PR DESCRIPTION
**Pull request by bzz**
- smaller size of input data per thread
 - fix env var name in docker compose
 - add more memory for Spark in local mode (docker compose)
 - switch to efficient serialization library
 - change caching strategy (spill on disk serialized objects)

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -6,9 +6,11 @@ services:
       - ./repositories:/repositories
       - ./query:/query
     environment:
-      DB_HOST: scylla
+      HOST: scylla
       BBLFSH_HOST: bblfshd
       FEATURES_EXTRACTOR_HOST: featurext
+    ports:
```
Comment (smacker) [inline docker-compose.yml]: one of the reasons why there are no ports - make sure user can run gemini without port collisions.

File: hash [hunk 2]
```diff
@@ -55,8 +56,12 @@ sparkSubmit \
   --master "${MASTER:=local[*]}" \
   --name "${app_name}" \
   --jars "${deps_jar}" \
+  --conf "spark.driver.memory=4g" \
```
Comment (smacker) [inline hash]: does it mean we require 4GB of memory to be given to docker?
If yes we need at least to document it. For "docker for mac" the default value is 1GB if I remember right.

File: src/main/scala/tech/sourced/gemini/Hash.scala [hunk 3]
```diff
@@ -33,8 +34,8 @@ class Hash(session: SparkSession, log: Slf4jLogger) {
     * @param repos DataFrame with engine.getRepositories schema
     */
   def forRepos(repos: DataFrame): HashResult = {
-    val files = filesForRepos(repos).cache()
-    val uasts = extractUast(files).cache()
+    val files = filesForRepos(repos).persist(StorageLevel.MEMORY_AND_DISK_SER)
+    val uasts = extractUast(files).persist(StorageLevel.MEMORY_AND_DISK_SER)
```
Comment (smacker) [inline src/main/scala/tech/sourced/gemini/Hash.scala]: should we make this option configurable?

File: hash [hunk 4]
```diff
@@ -55,8 +56,12 @@ sparkSubmit \
...
   --conf "spark.executor.memory=4g" \
   --conf "spark.default.parallelism=8" \
+  --conf "spark.serializer=org.apache.spark.serializer.KryoSerializer" \
```
Comment (smacker) [inline hash]: According to the [docs](https://spark.apache.org/docs/latest/tuning.html#data-serialization) KryoSerializer requires classes registration. Should we at least create an issue for that?

File: docker-compose.yml (hunk 1 above)
Comment (bzz) [reply docker-compose.yml]: I see. But without having this we would not be able to point users of `docker-compose` to Spark web ui, which is crucial for debugging.

File: hash (hunk 2 above)
Comment (bzz) [reply hash]: No, it does not. I just max capacity that the Driver JVM process possibly can take.

File: src/main/scala/tech/sourced/gemini/Hash.scala (hunk 3 above)
Comment (bzz) [reply src/main/scala/tech/sourced/gemini/Hash.scala]: We will, one day, after fine-tuning and understanding performance implications better. For now, it just changes the default persistence level from one to another.

File: hash (hunk 4 above)
Comment (bzz) [reply hash]: 👍 this is not mandatory but indeed will save some space later on https://github.com/src-d/gemini/issues/141

**Comment by bzz** [comment]
> the main concern is how gemini will work with default configuration on docker

Valid concern, thank you for raising it. In short - it will not. 

Long story: you can see it yourself - a [quick search](https://github.com/apache/spark/search?q=spark+driver+memory&unscoped_q=spark+driver+memory) over Spark source code reveals that [this option](https://github.com/apache/spark/blob/f5fe63f7b8546b0102d7bfaf3dde77379f58a4d1/launcher/src/main/java/org/apache/spark/launcher/SparkClassCommandBuilder.java#L111) will set JVM `-Xmx` argument that is a standard for setting the [maximum heap size](https://www.eclipse.org/openj9/docs/xms/) of the process.

Hope this addresses the concern.

File: docker-compose.yml (hunk 1 above)
Comment (smacker) [reply docker-compose.yml]: I would suggest such users use `.host.yaml` config. But okay. The only problem I'll have to edit `docker-compose.yml` every time when I want to test it with docker-compose but I can live with it.

**Pull request merged**